pkg/fleetautoscalers/controller.go

```diff
@@ -0,0 +1,363 @@
+	"github.com/mattbaird/jsonpatch"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	admv1beta1 "k8s.io/api/admission/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	extclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1beta1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+)
+
+// Controller is a the FleetAutoscaler controller
+type Controller struct {
+	logger                *logrus.Entry
+	crdGetter             v1beta1.CustomResourceDefinitionInterface
+	fleetGetter           getterv1alpha1.FleetsGetter
+	fleetLister           listerv1alpha1.FleetLister
+	fleetAutoscalerGetter getterv1alpha1.FleetAutoscalersGetter
+	fleetAutoscalerLister listerv1alpha1.FleetAutoscalerLister
+	fleetAutoscalerSynced cache.InformerSynced
+	workerqueue           *workerqueue.WorkerQueue
+	recorder              record.EventRecorder
+}
+
+// NewController returns a controller for a FleetAutoscaler
+func NewController(
+	wh *webhooks.WebHook,
+	health healthcheck.Handler,
+	kubeClient kubernetes.Interface,
+	extClient extclientset.Interface,
+	agonesClient versioned.Interface,
+	agonesInformerFactory externalversions.SharedInformerFactory) *Controller {
+
+	agonesInformer := agonesInformerFactory.Stable().V1alpha1()
+	fasInformer := agonesInformer.FleetAutoscalers().Informer()
+
+	c := &Controller{
+		crdGetter:             extClient.ApiextensionsV1beta1().CustomResourceDefinitions(),
+		fleetGetter:           agonesClient.StableV1alpha1(),
+		fleetLister:           agonesInformer.Fleets().Lister(),
+		fleetAutoscalerGetter: agonesClient.StableV1alpha1(),
+		fleetAutoscalerLister: agonesInformer.FleetAutoscalers().Lister(),
+		fleetAutoscalerSynced: fasInformer.HasSynced,
+	}
+	c.logger = runtime.NewLoggerWithType(c)
+	c.workerqueue = workerqueue.NewWorkerQueue(c.syncFleetAutoscaler, c.logger, stable.GroupName+".FleetAutoscalerController")
+	health.AddLivenessCheck("fleetautoscaler-workerqueue", healthcheck.Check(c.workerqueue.Healthy))
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(c.logger.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events("")})
+	c.recorder = eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "fleetautoscaler-controller"})
+
+	kind := stablev1alpha1.Kind("FleetAutoscaler")
+	wh.AddHandler("/mutate", kind, admv1beta1.Create, c.creationMutationHandler)
+	wh.AddHandler("/validate", kind, admv1beta1.Update, c.mutationValidationHandler)
+
+	fasInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
```

Comment:
Took this for a spin - works perfectly :+1: 

Question for you - this is only going to run about every ~30 seconds (which is what the current default sync is). Was wondering if that was by design?

I know we talked about slowing down the dropping of the of `GameServers` - but we could watch for the `Allocation` of `GameServers` (watch for `GameServer` update events, where the new state is `Allocated`), and enqueue the FAS at that point - then we would get realtime scaling up, rather than up to 30 seconds wait time.

I'm very happy to drop in the change myself once this PR is merged as well -- not a blocking issue.

Reply:
Actually yes, the current behavior is intended.

This is how autoscalers work afaik, they sample the current state at regular intervals.
This is done to reduce noise and to decouple the workload. In a live game , if we link the autoscaler to pod events it will work non-stop.

I also don't think this is really needed. The buffer size you set is supposed to cover the load between scaling intervals.

But sure, if this is really needed maybe it can be added as an optional flag in the spec.
It should come in a package with a cooldown value as well.
So yes, I think it should be done in a separate PR as well.